File: 122.买卖股票的最佳时机-ii.py
#
# @lc app=leetcode.cn id=122 lang=python3
#
# [122] 买卖股票的最佳时机 II
#

# @lc code=start

from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        if n < 2:
            return 0
        cash = 0
        stock = -prices[0]

        preCash = cash
        preStock = stock

        for price in prices[1:]:
            cash = max(preCash, preStock + price)
            stock = max(preCash - price, preStock)

            preCash = cash
            preStock = stock

        return cash


# @lc code=end

File: 122.买卖股票的最佳时机-ii.ts
/*
 * @lc app=leetcode.cn id=122 lang=typescript
 *
 * [122] 买卖股票的最佳时机 II
 */

// @lc code=start

const { max } = Math;

function maxProfit(prices: number[]): number {
    const n = prices.length;

    let dp_i_0 = 0,
        dp_i_1 = -prices[0];

    for (let i = 1; i < n; i += 1) {
        // 空仓，持仓
        [dp_i_0, dp_i_1] = [
            max(dp_i_0, dp_i_1 + prices[i]),
            max(dp_i_1, dp_i_0 - prices[i]),
        ];
    }
    return dp_i_0;
}

// @lc code=end
